tinygrad/tensor.py

```diff
@@ -617,6 +617,38 @@ def dropout(self, p=0.5) -> Tensor:
     mask = (Tensor.rand(*self.shape, requires_grad=False) >= p).cast(dtypes.bool)
     return self * mask * (1/(1.0 - p))
 
+  # note: the way onehot is implemented, if value is out of range, all 0s are returned
+  def onehot(self, num_classes) -> Tensor: 
+    assert dtypes.is_int(self.dtype), f"onehot only supported for ints, got {self.dtype}"
+    return self.reshape(list(self.shape)+[1]).repeat([1]*len(self.shape)+[num_classes]).eq(Tensor.arange(num_classes, dtype=self.dtype)) # could cast to smaller type, currently just same as self
+  
+  def negative_log_likelihood(self, target:Tensor, weight:Optional[Tensor] = None, ignore_index=-100, reduction:Literal['none', 'mean', 'sum'] = 'mean') -> Tensor: 
```

Comment:
This is tons of lines.

Reply:
yes agree, the extra asserts where to see if something unexpected was going on with the types, ref the failed test. But seems to maybe be HW thing.
Pushed more compact impl.

extra: though if the idea of making the type checks HW-agnostic by instead checking dtype priorities and assure that way the behavior is as expected. For example a future compact accelerator only having float16 should still work.

Reply:
the shape asserts is not needed but makes the error message more understandable, don't have to dig into the code